### PR TITLE
Feat/better mixer

### PR DESCRIPTION
**Pull request by nenikitov**

### CODE REVIEW

File: engine/src/asset/sound/dat/mixer.rs [hunk 1]
```diff
@@ -1,51 +1,541 @@
...
+                    envelope
+                        .volume_end()
+                        .get(
+                            self.pos_volume_envelope
+                                .saturating_sub(envelope.volume_beginning().len()),
+                        )
+                        .copied()
+                }
+            }
+            TInstrumentVolume::Constant(_) => Some(1.),
```
Comment (UserIsntAvailable) [inline engine/src/asset/sound/dat/mixer.rs:89]: The value inside `TInstrumentVolume::Constant` is always `1.0`...

```suggestion
            TInstrumentVolume::Constant(volume) => Some(volume),
```

**Review by UserIsntAvailable — CHANGES_REQUESTED** [review]
I might have missed something (way too many files to check), overall looks good.

File: engine/src/asset/sound/dat/mixer.rs [hunk 2]
```diff
@@ -1,51 +1,541 @@
...
+        if let Some((previous, position)) = &mut self.previous {
+            let factor = (*position / Self::SAMPLE_BLEND) as f32;
+            let previous_sample = previous.generate_sample(step);
+
+            *position += step;
+            if *position >= Self::SAMPLE_BLEND {
+                self.previous = None;
+            }
+
+            previous_sample + factor * (current_sample - previous_sample)
```
[the remark this answers is not in the document]
Comment (nenikitov) [reply engine/src/asset/sound/dat/mixer.rs]: Genuinely didn't know that this exists. Tyty

File: engine/src/asset/sound/dat/mixer.rs (hunk 1 above)
Comment (nenikitov) [reply engine/src/asset/sound/dat/mixer.rs]: Yeah no clue why I did it. I refactored it to an `Option` instead.

File: engine/src/asset/sound/dat/t_song.rs [hunk 3]
```diff
@@ -68,9 +106,18 @@ impl AssetParser<Wildcard> for TSong {
...
 
-            let (_, orders) = multi::count!(number::le_u8, header.song_length as usize)(
-                &input[pointers.orders as usize..],
-            )?;
+            let samples = uncompress(&input[pointers.sample_data as usize..]);
+            let (_, samples) = multi::count!(
+                TSample::parser(&samples),
+                header.sample_count as usize
+            )(&input[pointers.samples as usize..])?;
+            let samples = samples.into_iter().map(Rc::new).collect::<Vec<_>>();
```
[the remark this answers is not in the document]
Comment (nenikitov) [reply engine/src/asset/sound/dat/t_song.rs:114]: I didn't use `collect_vec` anywhere else and I'd prefer to stay consistent. For whatever reason I prefer `collect::<Vec<_>>`....

File: engine/src/asset/sound/sample.rs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    fn wave_le_bytes(&self) -> [u8; Self::SIZE_BYTES] {
+        self.to_le_bytes()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct AudioBuffer<S: AudioSamplePoint> {
+    pub data: Vec<S>,
```
[the remark this answers is not in the document]
Comment (nenikitov) [reply engine/src/asset/sound/sample.rs:55]: Idk about this one, ideally they shouldn't be `pub` but they need for `format` for outputting the results of the tests. I think I want to keep like this for now until we stabilize how the files should be output - ideally inside `main` and not as a result of a test.

**Review by UserIsntAvailable — APPROVED** [review]
Looks good; there is 1 place where you left a `0.`, but I can change it myself and just merge this :) :tada: :taco:

**Pull request merged**